sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1556,6 +1556,20 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       df.where($"city".contains(new java.lang.Character('A'))),
       Seq(Row("Amsterdam")))
   }
+
+  test("SPARK-25942: typed aggregation on primitive type") {
+    val ds = Seq(1, 2, 3).toDS()
+
+    val agg = ds.groupByKey(_ >= 2)
+      .agg(sum("value").as[Long], sum($"value" + 1).as[Long])
+    assert(agg.collect() === Seq((false, 1, 2), (true, 5, 7)))
+  }
+
+  test("SPARK-25942: typed aggregation on product type") {
+    val ds = Seq((1, 2), (2, 3), (3, 4)).toDS()
+    val agg = ds.groupByKey(x => x).agg(sum("_1").as[Long], sum($"_2" + 1).as[Long])
+    assert(agg.collect().sorted === Seq(((1, 2), 1, 3), ((2, 3), 2, 4), ((3, 4), 3, 5)))
```

Comment:
can we use `checkAnswer`/`CheckDataset`?

Reply:
Using `checkDataset` comes out an error:
```
[error]  found   : org.apache.spark.sql.Dataset[((Int, Int), Long, Long)]
[error]  required: org.apache.spark.sql.Dataset[((Int, Int), AnyVal, AnyVal)]
[error] Note: ((Int, Int), Long, Long) <: ((Int, Int), AnyVal, AnyVal), but class Dataset is invariant in type T.
[error] You may wish to define T as +T instead. (SLS 4.5)
[error]     checkDataset(agg, ((1, 2), 1, 3), ((2, 3), 2, 4), ((3, 4), 3, 5))
```

Reply:
Is there any suggestion?

Reply:
can you try `((1, 2), 1L, 3L)` instead of `((1, 2), 1, 3)`?

Reply:
ah, cool, I see.